src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java

```diff
@@ -905,14 +960,111 @@ protected void createFlattenedDependencies( Model effectiveModel, List<Dependenc
         }
     }
 
+    private void createFlattenedDependenciesAll( List<Dependency> projectDependencies, List<Dependency> flattenedDependencies )
+            throws DependencyTreeBuilderException, ArtifactDescriptorException {
+        final Stack<DependencyNode> dependencyNodeStack = new Stack<>();
+        final Set<String> processedDependencies = new HashSet<>();
+
+        final Artifact projectArtifact = this.project.getArtifact();
+
+        final DependencyNode dependencyNode = this.dependencyTreeBuilder.buildDependencyTree(this.project,
+                this.localRepository, null);
+
+        dependencyNode.accept(new DependencyNodeVisitor() {
+            @Override public boolean visit(DependencyNode node) {
+                if (node.getArtifact().getGroupId().equals(projectArtifact.getGroupId()) && node.getArtifact().getArtifactId().equals(projectArtifact.getArtifactId())) {
+                    return true;
+                }
+                if (node.getState() != DependencyNode.INCLUDED) return true;
+                dependencyNodeStack.push(node);
+                return true;
+            }
+
+            @Override public boolean endVisit(DependencyNode node) {
+                return true;
+            }
+        });
+
+        while (!dependencyNodeStack.isEmpty()) {
+            DependencyNode node = dependencyNodeStack.pop();
+
+            Artifact artifact = node.getArtifact();
+
+            Dependency dependency = new Dependency();
+            dependency.setGroupId(artifact.getGroupId());
+            dependency.setArtifactId(artifact.getArtifactId());
+            dependency.setVersion(artifact.getVersion());
+            dependency.setClassifier(artifact.getClassifier());
+            dependency.setOptional(artifact.isOptional());
+            dependency.setScope(artifact.getScope());
+            dependency.setType(artifact.getType());
+
```

Comment:
Would you add comment on why exclusion helps here?

Reply:
added to comment.